### PR TITLE
Fix read_sql empty result with chunksize bug GH34411

### LINKED ISSUE #34411: BUG: pd.read_sql returns empty list if query has no results and chunksize is set
- [ x] I have checked that this issue has not already been reported.

- [ x] I have confirmed this bug exists on the latest version of pandas.

- [x ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import sqlite3

# Create empty test table in memory
conn = sqlite3.connect(':memory:')
conn.cursor().execute('CREATE TABLE test (column_1 INTEGER);')

# Run the query without chunksize, works as expected
pd.read_sql('select * from test', conn)

# Run the query with chunksize, returns generator as expected
pd.read_sql('select * from test', conn, chunksize=5)

# However, the generator is empty
list(pd.read_sql('select * from test', conn, chunksize=5))

# I would expect, that for all cases where chunksize isn't necessary,
# than the following two lines would return exactly the same
# result, but the second throws "ValueError: No objects to concatenate"
pd.read_sql('select * from test', conn)

pd.concat(pd.read_sql('select * from test', conn, chunksize=5))
```

#### Problem description

In many cases, returning zero rows is an expected result, and the code should run fine on the returned DataFrame (iterating over it, getting all values in a row, etc.).

The current behaviour instead returns an empty list, with no information about for example the columns in the dataframe.

#### Expected Output

The expected output would be a list containing a single empty dataframe, with the correct column metadata. I would expect that for all queries that run fine without chunkbite being set, the following equality should hold:

```
pd.testing.assert_frame_equal(
    pd.read_sql(query, conn),
    pd.concat(pd.read_sql(query, conn, chunksize=5))
)
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 333db4b765f8e88c0c2392943cb7d6c6013dc6e8
python           : 3.8.2.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
Version          : Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.1.0.dev0+1685.g333db4b76.dirty
numpy            : 1.18.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 46.4.0.post20200518
Cython           : 0.29.19
pytest           : 5.4.2
hypothesis       : 5.16.0
sphinx           : 3.0.4
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.14.0
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : 1.3.2
fastparquet      : 0.4.0
gcsfs            : None
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : 0.17.1
pytables         : None
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.4.1
sqlalchemy       : 1.3.17
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.15.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.48.0

</details>

### PR DESCRIPTION
**Pull request by JohanKahrstrom**
- [x ] closes #34411
- [ x] tests added / passed
- [ x] passes `black pandas`
- [ x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ x] whatsnew entry

`pd.read_sql` was returning an empty generator when chunksize is set and the query returns zero results. Now it correctly returns a generator with a single empty DataFrame (:issue:`34411`).

### CODE REVIEW

File: pandas/io/sql.py [hunk 1]
```diff
@@ -837,11 +837,17 @@ def _query_iterator(
...
     ):
         """Return generator through chunked result set."""
+        has_read_data = False
         while True:
             data = result.fetchmany(chunksize)
             if not data:
+                if not has_read_data:
+                    yield DataFrame.from_records(
+                        [], columns=columns, coerce_float=coerce_float
+                    )
```
Comment (dsaxton) [inline pandas/io/sql.py:847]: What happens if you just yield an empty DataFrame here without keeping track of having read data?

Comment (JohanKahrstrom) [reply pandas/io/sql.py:847]: If the query does return rows, the resulting generator will have an empty dangling dataframe. Say the chunksize is set to 5, and the result set has 13 rows, the generator will contain:
```
(
   a DataFrame with 5 rows,
   a DataFrame with 5 rows,
   a DataFrame with 3 rows,
   an empty DataFrame
)
```

This is probably not desirable. I agree, keeping track of if we read data isn't very elegant, but I it's the best solution I can come up with.

File: pandas/io/sql.py [hunk 2]
```diff
@@ -837,11 +837,17 @@ def _query_iterator(
...
                 break
```
Comment (WillAyd) [inline pandas/io/sql.py:848]: Is this `break` statement still being hit?

Comment (JohanKahrstrom) [reply pandas/io/sql.py:848]: Yes, if `has_read_data` an empty dataframe is yielded, then the `break` statement breaks out of the `while True` loop, ending the generator.

You can verify that this statement is hit by running for example

```
> pytest --cov=pandas/io/ --cov-report html pandas/tests/io/test_sql.py::TestSQLApi::test_read_sql_with_chunksize_no_result
```

**Review by WillAyd — APPROVED** [review]
lgtm @jorisvandenbossche

**Comment by WillAyd** [comment]
@JohanKahrstrom looks like a merge conflict - can you fix that up? ping @jorisvandenbossche or @jreback for second review

**Comment by JohanKahrstrom** [comment]
@WillAyd, @jorisvandenbossche, @jreback, Merge conflict has been resolved, and commit history squashed.

**Comment by JohanKahrstrom** [comment]
... and fixed black formatting.

**Comment by JohanKahrstrom** [comment]
Thanks Will, I was trying to figure out how to rerun the azure check...

File: pandas/io/sql.py [hunk 3]
```diff
@@ -1684,14 +1700,20 @@ def _query_iterator(
         cursor, chunksize, columns, index_col=None, coerce_float=True, parse_dates=None
     ):
         """Return generator through chunked result set"""
+        has_read_data = False
```
Comment (jreback) [inline pandas/io/sql.py:1703]: you could try defining data in an out scope e.g.
```
data: List
```

then you can test that after the break

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/sql.py (hunk 3 above)
Comment (JohanKahrstrom) [reply pandas/io/sql.py:1703]: Thank you Jeff.

If I understand what you're proposing correctly (i.e. defining `data` outside the while loop then checking it's status after exiting the loop), then that unfortunately doesn't work. The break clause is triggered when `not data` is True (i.e. `data` is an empty list), so regardless of how many results are returned (and how many times the loop is executed) when you exit the loop `data` will be an empty list. Line 1705 `data = cursor.fetchmany(chunksize)` is always executed at least once, so setting `data` to `None` (or anything else) outside the while loop doesn't achieve anything.

**Comment by jreback** [comment]
can u merge master and we will look again

**Comment by simonjayhawkins** [comment]
@JohanKahrstrom can you add a release note

**Comment by JohanKahrstrom** [comment]
@jreback @simonjayhawkins Added release note, merged with master. Unrelated checks are failing, not sure what to do about those...

**Comment by dsaxton** [comment]
@JohanKahrstrom Can you merge master to fix conflicts? Looks like this might be close to ready?

**Comment by JohanKahrstrom** [comment]
@dsaxton I've merged with master. There's 3 failing tests unrelated to my code change, and the Azure test suite failed again (I don't have sufficient privileges to re-run them).

@jreback are you happy with my argument for not implementing your suggested change (as far as I can see it won't work)?

**Comment by arw2019** [comment]
@JohanKahrstrom can you merge master once more? (There's a conflict in whatsnew)

**Comment by JohanKahrstrom** [comment]
@arw2019 Done.

**Comment by arw2019** [comment]
This looks ready to go.

cc @jreback @dsaxton

**Review by jreback — CHANGES_REQUESTED** [review]
@JohanKahrstrom can you merge master and move the whatsnew note. i think this was ok.

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -669,6 +669,7 @@ I/O
 - :func:`read_csv` was closing user-provided binary file handles when ``engine="c"`` and an ``encoding`` was requested (:issue:`36980`)
 - Bug in :meth:`DataFrame.to_hdf` was not dropping missing rows with ``dropna=True`` (:issue:`35719`)
 - Bug in :func:`read_html` was raising a ``TypeError`` when supplying a ``pathlib.Path`` argument to the ``io`` parameter (:issue:`37705`)
+- :meth:`read_sql` returned an empty generator if ``chunksize`` was no-zero and the query returned no results. Now returns a generator with a single empty dataframe (:issue:`34411`)
```
[the remark this answers is not in the document]
Comment (JohanKahrstrom) [reply doc/source/whatsnew/v1.2.0.rst]: @jreback Done, thanks.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @JohanKahrstrom

**Pull request merged**